### ISSUE TITLE
[$250] Distance - Unable to save distance when Start and Stop waypoints are valid address

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#48401]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v9.0.27-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4908819
**Email or phone of affected tester (no customers):** applausetester+kh050806@applause.expensifail.com
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-Internal team

## Action Performed:
1. Go to staging.new.expensify.com
2. Go to workspace chat.
3. Submit a distance expense.
4. Go to transaction thread.
5. Go offline.
6. Click Distance.
7. Change both Start and Stop waypoints to invalid address ( use random strings).
8. Save it.
9. Go online.
10. Click on the header subtitle to return to main chat.
11. Click on the expense preview.
12. Click Distance.
13. Note that the waypoints are valid addresses (from Step 3).
14. Click Save.

## Expected Result:
The distance will be saved because Start and Stop waypoints are valid.

## Actual Result:
Nothing happens after clicking Save button when Start and Stop waypoints are valid.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/bddd22db-5784-490e-8c3f-076b2da15e1a

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~013c32d2d6ff5c2253</li>
        <li>Upwork Job ID: 1830768674764828148</li>
        <li>Last Price Increase: 2024-11-13</li>
<li>Automatic offers: </li>
<ul>
<li>wildan-m | Contributor | 104956787</li>
</ul></li>
    </ul>
</details>

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Nothing happens after clicking Save button when Start and Stop waypoints are valid.

### What is the root cause of that problem?
After we go to the expense report or transaction thread report again, the transaction data is returned from BE that
 will reset the waypoint data of the transaction but the transaction error isn't cleared. Then we cannot save the distance without changing it.

https://github.com/Expensify/App/blob/ec8a5d974d18dea142da2c27e4a8a27ce0a70794/src/pages/iou/request/step/IOURequestStepDistance.tsx#L100

https://github.com/Expensify/App/blob/ec8a5d974d18dea142da2c27e4a8a27ce0a70794/src/pages/iou/request/step/IOURequestStepDistance.tsx#L398-L400
### What changes do you think we should make in order to solve the problem?
In `MoneyRequestView`, we can create a `useEffect` that will clear the route error if `hasRoute` is `true` and the route transaction error still exists. That will ensure that the route error is only cleared when the route is re-fetched because we can have a case after the waypoint error appears user can offline and open the expense report/transaction thread report again (in this case, the waypoint isn't re-fetched and we don't want to clear the error). We can create a new util to only clear the route error or re-use `Transaction.clearError` function.

```
useEffect(() => {
    if (!transaction?.errorFields?.route || !hasRoute) {
        return;
    }

    Transaction.clearRouteError(transaction.transactionID);
}, [hasRoute, transaction?.errorFields?.route, transaction?.transactionID]);
```


### What alternative solutions did you explore? (Optional)
NA

**Comment by sobitneupane** [comment]
@Krishna2323 @nkdengineer Looks like the issue is not reproducible any longer.

I cannot reproduce the issue.

**Comment by nkdengineer** [comment]
@sobitneupane I can still able to reproduce this bug on the latest main.

https://github.com/user-attachments/assets/dc4aef04-a034-4ca6-9f47-5a679437bacd

**Comment by sobitneupane** [comment]
Thanks @nkdengineer 

I don't think we should remove the error. The invalid waypoint error will be shown in the transaction thread, and when the user opens the transaction, they won't find any error. We should display the error alongside the invalid addresses.

**Comment by sobitneupane** [comment]
> What are the different cases that can lead to a route error (apart from an invalid waypoint)?

I can think of these two cases:
- no route found
- Route exceeds max distance allocation.

```
    const hasRouteError = !!transaction?.errorFields?.route;
    const isEditing = action === CONST.IOU.ACTION.EDIT; 
    const [optimisticWaypoints, setOptimisticWaypoints] = useState<WaypointCollection | null>(hasRouteError && isEditing ? transactionBackup?.modifiedWaypoints ?? null : null);  
```

@Krishna2323 It does not solve the issue. I can still reproduce the issue.


https://github.com/user-attachments/assets/bedd7178-b133-4e53-a118-cb3f7cfde7af

**Comment by nkdengineer** [comment]
> I don't think we should remove the error. The invalid waypoint error will be shown in the transaction thread, and when the user opens the transaction, they won't find any error. We should display the error alongside the invalid addresses.

@sobitneupane The problem here is after we re-open the expense report, BE will return the data that overrides the invalid waypoint.

**Comment by sobitneupane** [comment]
> The problem here is after we re-open the expense report, BE will return the data that overrides the invalid waypoint.

@nkdengineer I got it. If we don't show an error, the user might think their change was successful.

**Comment by sobitneupane** [comment]
Waiting for proposal.

**Comment by wildan-m** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Distance - Unable to save distance when Start and Stop waypoints are valid addresses

### What is the root cause of that problem?

Server send error and this check will be true:

https://github.com/Expensify/App/blob/7af7a3f85e040be3b862fe4a6dfeeeeccffed8aa/src/pages/iou/request/step/IOURequestStepDistance.tsx#L408-L411

### What changes do you think we should make in order to solve the problem?

We should re-validate the route on the initialMount, since the error remains, but waypoints has corrected from api `failureData`

In `src/hooks/useFetchRoute.ts` add new state `isInitialMount` with default to true

```tsx
    const [isInitialMount, setIsInitialMount] = useState(true);
```

And change `shouldFetchRoute` logic

https://github.com/Expensify/App/blob/7af7a3f85e040be3b862fe4a6dfeeeeccffed8aa/src/hooks/useFetchRoute.ts#L23

To

```tsx
   const shouldFetchRoute = (isDistanceRequest && (isRouteAbsentWithoutErrors || haveValidatedWaypointsChanged || isInitialMount) && !isLoadingRoute && Object.keys(validatedWaypoints).length > 1);
```

[Branch for this solution](https://github.com/Expensify/App/compare/main...wildan-m:wildan/fix/48401-unable-save-waypoint-on-error?expand=1)
### What alternative solutions did you explore? (Optional)
N/A

**Comment by wildan-m** [comment]
[Proposal updated](https://github.com/Expensify/App/issues/48401#issuecomment-2362722095)

- or only initial re-validate if it has route error

**Comment by wildan-m** [comment]
[Proposal Updated](https://github.com/Expensify/App/issues/48401#issuecomment-2362722095)
- handle issue when offline -> online, error shown -> offline

**Comment by sobitneupane** [comment]
Thanks for the proposal @wildan-m  Can you please explain the output after the change?

> And we need to skip hasRouteError validation when submitWaypoints in offline mode:
src/pages/iou/request/step/IOURequestStepDistance.tsx

Can you please add reasoning for this change?


In my understanding, the issue occurs because, instead of displaying the invalid addresses when opening the distance page, we are showing a previous valid route along with errors from the invalid route, which leads to the problem. I believe [this](https://github.com/Expensify/App/issues/48401#issuecomment-2338167468) should be the expected output.

**Comment by wildan-m** [comment]
[Proposal updated](https://github.com/Expensify/App/issues/48401#issuecomment-2362722095)

- Add reasoning behind allowing save button with unvalidated waypoints in offline
- Add alternative solution


> In my understanding, the issue occurs because, instead of displaying the invalid addresses when opening the distance page, we are showing a previous valid route along with errors from the invalid route, which leads to the problem. I believe https://github.com/Expensify/App/issues/48401#issuecomment-2338167468 should be the expected output.

That's true, but when we call the API, the `failedData` section of the API will restore the value to its previous correct value.

There is no guarantee that the user will re-open that distance request, restoring it to previous correct value is better than keeping the wrong values.

Consider this scenario:
If we keep the wrong values, then user clicking the X button to clear the error, which function will fix the wrong values? assume the user will not re-open the distance page and stay on its parent report.

**Comment by sobitneupane** [comment]
@kadiealexander When submitting a distance expense(offline) with invalid waypoints (random strings for Start and Stop addresses), an error is shown and the invalid waypoints are optimistically saved. However, after coming back online and returning to the transaction thread shows the valid waypoints, as fetched from the API, instead of the invalid ones. The issue arises because the transaction details are re-fetched, which updates the waypoints to valid ones but retains the error message from the original invalid waypoints. This creates a mismatch where valid waypoints are displayed, but the error still reflects the previous invalid input. The question is: should we display the invalid waypoints along with the error message, or should we show the valid waypoints returned by the API and remove the error?

**Comment by sobitneupane** [comment]
Started discussion in slack: https://expensify.slack.com/archives/C01GTK53T8Q/p1727685264166239

**Comment by kadiealexander** [comment]
@sobitneupane in my opinion, we should show the valid waypoints returned by the API and remove the error.

**Comment by sobitneupane** [comment]
Thanks for the input, @kadiealexander.

I will re-review the proposals, taking [this comment](https://github.com/Expensify/App/issues/48401#issuecomment-2387658858) into account as the expected output.

**Comment by sobitneupane** [comment]
Thanks for the proposal, @wildan-m 

As suggested [here](https://github.com/Expensify/App/issues/48401#issuecomment-2387658858), we should remove the error as well.

**Comment by sobitneupane** [comment]
> Or we can do that in IOURequestStepDistance. We can check if there is no error except the route error and then we can clear the errors.

@Krishna2323 Can you please share more details on how you intend to implement the plan.

**Comment by wildan-m** [comment]
> As suggested https://github.com/Expensify/App/issues/48401#issuecomment-2387658858, we should remove the error as well.

@sobitneupane my solution will remove the error once the server validated the waypoints.

Why the error restored when we navigate back? that's because we aren't updating the transactionBackup waypoints data when we are online so error persists in backup and will be restored when we navigate back from waypoint editor. This navigate back issue will be resolved after the PR in this issue merged https://github.com/Expensify/App/issues/48630

**Comment by sobitneupane** [comment]
@wildan-m In that case we should wait for https://github.com/Expensify/App/issues/48630 to get resolved.

**Comment by sobitneupane** [comment]
Still waiting on https://github.com/Expensify/App/issues/48630

**Comment by wildan-m** [comment]
[Proposal Updated](https://github.com/Expensify/App/issues/48401#issuecomment-2362722095)

- Merge branch with the latest main that include PR for this issue https://github.com/Expensify/App/issues/48630
- Remove unnecessary change so the waypoints not undefined

@sobitneupane, I've tested it and the error message removed as expected.

**Comment by sobitneupane** [comment]
@wildan-m I can still reproduce the error with the changes you suggested. Can you please retest your branch?

**Comment by wildan-m** [comment]
@sobitneupane Ah, forgot to mention that the GET_ROUTE_FOR_BACKUP endpoint not exists in dev, it's in staging and prod, please point your server to staging in troubleshoot page. Ref: https://github.com/Expensify/App/issues/48630#issuecomment-2437329298


https://github.com/user-attachments/assets/f7a7bfeb-f9b3-4420-aab9-2826add8ef8f

**Comment by kadiealexander** [comment]
@sobitneupane have you had a chance to retest?

**Comment by sobitneupane** [comment]
> And remove transactionBackup?.pendingFields?.waypoints here so it can contain values for the case in this issue
```diff
-    const backupWaypoints = transactionBackup?.pendingFields?.waypoints ? transactionBackup?.comment?.waypoints : undefined;
+    const backupWaypoints = transactionBackup?.comment?.waypoints;
```
@wildan-m Could you please explain this part of your proposal? I'm not entirely sure why this change is necessary. Additionally, could you point to the original proposal that introduced this change.

**Comment by wildan-m** [comment]
@sobitneupane I authored the change in https://github.com/Expensify/App/pull/51519 to address the issue in https://github.com/Expensify/App/issues/48630#issuecomment-2365034676.

Why did we add that check in the first place?

The issue is with pending distance requests, so it makes sense to limit the backupWaypoint when it's pending for that specific case.

Then come the issue in this post, I didn't consider the case for this issue when creating the PR, but after get back to this issue and re-testing my solution, I've noticed that we also need `backupWaypoints` when it's not pending, without that, `useFetchRoute` for backup transaction will not trigger because inputted waypoints is empty.

In summary, with or without `transactionBackup?.pendingFields?.waypoints` check, https://github.com/Expensify/App/issues/48630 will still work and we need to remove that check so the backupTransaction for this issue also updated when it's updated, the error will be removed when the waypoints corrected

**Comment by sobitneupane** [comment]
Thanks for the update @wildan-m 

Let's make sure to test and add videos for https://github.com/Expensify/App/issues/48630 issue as well in the PR.

[Proposal](https://github.com/Expensify/App/issues/48401#issuecomment-2362722095) from @wildan-m looks good to me.

🎀 👀 🎀 C+ reviewed

**Comment by wildan-m** [comment]
@sobitneupane The PR is ready https://github.com/Expensify/App/pull/52736. Thanks!

**Comment by sobitneupane** [comment]
PR in review.

**Comment by wildan-m** [comment]
@yuwenmemon @kadiealexander We came across [this issue](https://github.com/Expensify/App/pull/52736#issuecomment-2618624667) during testing. Could you assist in checking the `GetRoute` and `GetRouteForBackup` APIs to see if they will clear the error when valid waypoints are submitted, or they only return coordinate data?

The error cleanup for the relevant transaction may look like this.

```tsx
                    errorFields: {
                        route: null,
                        routes: null,
                        waypoints: null,
                    },
```

**Comment by wildan-m** [comment]
Asking in slack
https://expensify.slack.com/archives/C01GTK53T8Q/p1738899434114649
